cas/service.c

```diff
@@ -1082,6 +1082,31 @@ static int op_sync_wait(struct m0_fom *fom)
 	return M0_FSO_AGAIN;
 }
 
+/**
+ * Check specific deadlock case for PUT and DEL operation
+ * 1. PUT - lock->tx_open
+ * 2. DEL - tx_open->lock
+ */
+M0_INTERNAL bool m0_cas_fom_in_deadlock(const struct m0_fom *fom0)
+{
+	struct cas_fom         *fom   = M0_AMB(fom, fom0, cf_fom);;
+	struct m0_long_lock    *lock  = m0_ctg_lock(fom->cf_ctg);
+	const struct m0_be_tx  *tx    = &fom0->fo_tx.tx_betx;
+	struct m0_be_tx_group  *gr    = tx->t_group;
+	enum m0_cas_opcode      opc   = m0_cas_opcode(fom0->fo_fop);
+	int                     phase = m0_fom_phase(fom0);
+
+	if (gr->tg_state == M0_BGS_FROZEN && !cas_is_ro(opc) &&
+	    ((phase == M0_FOPH_TXN_WAIT &&
+	      m0_be_tx_state(tx) == M0_BTS_GROUPING &&
+	      m0_long_is_write_locked(lock, fom0)) ||
```

Comment:
[* doc *] Please explain in a comment why the use of this function is acceptable here.

In general the return value in such kind of functions is outdated even before the function returns.